cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceService.java

```diff
@@ -0,0 +1,215 @@
+   */
+  private void ensureDefaultNamespaceExists() {
+    ListenableFuture<Void> listenableFuture = asyncCreateDefaultNamespace();
+    Futures.addCallback(listenableFuture, new FutureCallback<Void>() {
+      @Override
+      public void onSuccess(Void aVoid) {
+        // No-op, since either default namespace already existed, or its creation was successful
+        LOG.info("Successfully ensured existence of default namespace.");
+      }
+
+      @Override
+      public void onFailure(Throwable throwable) {
+        LOG.warn("Failed to ensure existence of default namespace.");
+        // now keep retrying
+        if (isRunning()) {
+          final FutureCallback<Void> callback = this;
+          // Retry in 2 seconds. Shouldn't sleep in this callback thread. Should start a new thread for the retry.
+          Thread retryThread = new Thread("retrying-default-namespace-ensurer") {
+            @Override
+            public void run() {
+              try {
+                TimeUnit.SECONDS.sleep(2);
+                LOG.info("Retrying to ensure that default namespace exists.");
+                Futures.addCallback(asyncCreateDefaultNamespace(), callback);
+              } catch (InterruptedException e) {
+                LOG.warn("Default namespace ensurer retry thread interrupted.");
+              }
+            }
+          };
+          retryThread.setDaemon(true);
+          retryThread.start();
+        }
+      }
+    });
+  }
+
+  /**
+   * Asynchronously calls #createDefaultNamespace to create the default namespace
+   */
+  private ListenableFuture<Void> asyncCreateDefaultNamespace() {
+    ListeningExecutorService executorService = MoreExecutors.listeningDecorator(
+      Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("default-namespace-ensurer")));
+    return executorService.submit(new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        createDefaultNamespace();
+        return null;
+      }
+    });
+  }
+  
+  /**
+   * Creates the default namespace at a more deterministic time.
+   * It should be removed once we stop support for v2 APIs, since 'default' namespace is only reserved for v2 APIs.
+   */
+  private void createDefaultNamespace() {
+    NamespaceMeta.Builder builder = new NamespaceMeta.Builder();
+    NamespaceMeta defaultNamespace = builder.setId(Constants.DEFAULT_NAMESPACE)
+      .setName(Constants.DEFAULT_NAMESPACE)
+      .setDescription(Constants.DEFAULT_NAMESPACE)
+      .build();
+
+    try {
+      createNamespace(defaultNamespace);
+      LOG.info("Successfully created 'default' namespace.");
+    } catch (AlreadyExistsException e) {
+      LOG.info("'default' namespace already exists.");
+    }
+  }
+
+  /**
+   * Lists all namespaces
+   *
+   * @return a list of {@link NamespaceMeta} for all namespaces
+   */
+  public List<NamespaceMeta> listNamespaces() {
+    startAndWait();
```

Comment:
In ``AppFabricServer``, we do not block until the ``NamespaceService`` is up. We just call ``start()``, so ``AppFabricServer`` does not wait to start other services. Hence, as a precaution in ``NamespaceService``, ``startAndWait()`` ensures that the service is completely up. Maybe it would be more clear if I do
```
if (!isRunning()) {
  startAndWait();
}
```
However, ``startAndWait()`` only starts if necessary ([doc](http://docs.guava-libraries.googlecode.com/git-history/v14.0/javadoc/com/google/common/util/concurrent/Service.html#startAndWait()), so it would be the same effect. This way I feel is less code, so preferred it.